File: messaging-engine/Sources/MessagingEngine/Protocol.swift
import Foundation

/// Protocol frame types for TCP communication
public enum FrameType: UInt8, Codable {
    case message = 0x01
    case ack = 0x02
    case heartbeat = 0x03
    case subscribe = 0x04
    case unsubscribe = 0x05
    case error = 0xFF
}

/// Protocol frame for network transmission
public struct ProtocolFrame: Codable {
    public let type: FrameType
    public let payload: Data
    public let sequenceNumber: UInt64
    public let timestamp: Date
    
    public init(type: FrameType, payload: Data, sequenceNumber: UInt64, timestamp: Date = Date()) {
        self.type = type
        self.payload = payload
        self.sequenceNumber = sequenceNumber
        self.timestamp = timestamp
    }
}

/// Protocol encoder/decoder for binary frames
public class ProtocolCodec {
    private static let headerSize = 25 // 1 (type) + 8 (seq) + 8 (timestamp) + 8 (payload size)
    
    /// Encode a frame to binary data
    public static func encode(_ frame: ProtocolFrame) throws -> Data {
        var data = Data()
        
        // Frame type (1 byte)
        data.append(frame.type.rawValue)
        
        // Sequence number (8 bytes)
        data.append(contentsOf: withUnsafeBytes(of: frame.sequenceNumber.bigEndian) { Data($0) })
        
        // Timestamp (8 bytes - Unix timestamp)
        let timestamp = UInt64(frame.timestamp.timeIntervalSince1970 * 1000) // milliseconds
        data.append(contentsOf: withUnsafeBytes(of: timestamp.bigEndian) { Data($0) })
        
        // Payload size (8 bytes)
        let payloadSize = UInt64(frame.payload.count)
        data.append(contentsOf: withUnsafeBytes(of: payloadSize.bigEndian) { Data($0) })
        
        // Payload
        data.append(frame.payload)
        
        return data
    }
    
    /// Decode binary data to a frame
    public static func decode(_ data: Data) throws -> ProtocolFrame {
        guard data.count >= headerSize else {
            throw ProtocolError.invalidFrame("Frame too short")
        }
        
        var offset = 0
        
        // Frame type
        guard let type = FrameType(rawValue: data[offset]) else {
            throw ProtocolError.invalidFrame("Invalid frame type")
        }
        offset += 1
        
        // Sequence number
        let seqData = data.subdata(in: offset..<offset+8)
        let sequenceNumber = seqData.withUnsafeBytes { $0.load(as: UInt64.self).bigEndian }
        offset += 8
        
        // Timestamp
        let timestampData = data.subdata(in: offset..<offset+8)
        let timestampMs = timestampData.withUnsafeBytes { $0.load(as: UInt64.self).bigEndian }
        let timestamp = Date(timeIntervalSince1970: Double(timestampMs) / 1000.0)
        offset += 8
        
        // Payload size
        let payloadSizeData = data.subdata(in: offset..<offset+8)
        let payloadSize = payloadSizeData.withUnsafeBytes { $0.load(as: UInt64.self).bigEndian }
        offset += 8
        
        // Payload
        guard data.count >= offset + Int(payloadSize) else {
            throw ProtocolError.invalidFrame("Payload incomplete")
        }
        let payload = data.subdata(in: offset..<offset+Int(payloadSize))
        
        return ProtocolFrame(
            type: type,
            payload: payload,
            sequenceNumber: sequenceNumber,
            timestamp: timestamp
        )
    }
}

/// Protocol errors
public enum ProtocolError: Error {
    case invalidFrame(String)
    case encodingError(String)
    case decodingError(String)
    case connectionError(String)
}

File: integration/messaging-client/protocol.go
package messaging

import (
	"encoding/binary"
	"errors"
	"time"
)

// FrameType represents the type of protocol frame
type FrameType uint8

const (
	FrameTypeMessage    FrameType = 0x01
	FrameTypeAck        FrameType = 0x02
	FrameTypeHeartbeat  FrameType = 0x03
	FrameTypeSubscribe  FrameType = 0x04
	FrameTypeUnsubscribe FrameType = 0x05
	FrameTypeError      FrameType = 0xFF
)

// ProtocolFrame represents a frame in the messaging protocol
type ProtocolFrame struct {
	Type          FrameType
	Payload       []byte
	SequenceNumber uint64
	Timestamp     time.Time
}

// EncodeFrame encodes a ProtocolFrame to binary data
func EncodeFrame(frame ProtocolFrame) ([]byte, error) {
	data := make([]byte, 0, 25+len(frame.Payload))
	
	// Frame type (1 byte)
	data = append(data, byte(frame.Type))
	
	// Sequence number (8 bytes, big endian)
	seqBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(seqBytes, frame.SequenceNumber)
	data = append(data, seqBytes...)
	
	// Timestamp (8 bytes, milliseconds since epoch, big endian)
	timestampMs := uint64(frame.Timestamp.UnixMilli())
	timestampBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(timestampBytes, timestampMs)
	data = append(data, timestampBytes...)
	
	// Payload size (8 bytes, big endian)
	payloadSize := uint64(len(frame.Payload))
	payloadSizeBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(payloadSizeBytes, payloadSize)
	data = append(data, payloadSizeBytes...)
	
	// Payload
	data = append(data, frame.Payload...)
	
	return data, nil
}

// DecodeFrame decodes binary data to a ProtocolFrame
func DecodeFrame(data []byte) (ProtocolFrame, error) {
	const headerSize = 25
	
	if len(data) < headerSize {
		return ProtocolFrame{}, errors.New("frame too short")
	}
	
	var frame ProtocolFrame
	var offset int
	
	// Frame type
	frame.Type = FrameType(data[offset])
	offset++
	
	// Sequence number
	frame.SequenceNumber = binary.BigEndian.Uint64(data[offset : offset+8])
	offset += 8
	
	// Timestamp
	timestampMs := binary.BigEndian.Uint64(data[offset : offset+8])
	frame.Timestamp = time.UnixMilli(int64(timestampMs))
	offset += 8
	
	// Payload size
	payloadSize := binary.BigEndian.Uint64(data[offset : offset+8])
	offset += 8
	
	// Payload
	if len(data) < offset+int(payloadSize) {
		return ProtocolFrame{}, errors.New("payload incomplete")
	}
	frame.Payload = make([]byte, payloadSize)
	copy(frame.Payload, data[offset:offset+int(payloadSize)])
	
	return frame, nil
}

